skrifa/src/scale/colr/instance.rs

```diff
@@ -0,0 +1,653 @@
+    colr: Colr<'a>,
+    index_map: Option<DeltaSetIndexMap<'a>>,
+    var_store: Option<ItemVariationStore<'a>>,
+    coords: &'a [F2Dot14],
+}
+
+impl<'a> ColrInstance<'a> {
+    /// Creates a new instance for the given `COLR` table and normalized variation
+    /// coordinates.
+    pub fn new(colr: Colr<'a>, coords: &'a [F2Dot14]) -> Self {
+        let index_map = colr.var_index_map().and_then(|res| res.ok());
+        let var_store = colr.item_variation_store().and_then(|res| res.ok());
+        Self {
+            colr,
+            coords,
+            index_map,
+            var_store,
+        }
+    }
+
+    /// Computes a sequence of N variation deltas starting at the given
+    /// `var_base` index.
+    fn var_deltas<const N: usize>(&self, var_index_base: u32, tys: [DeltaTy; N]) -> [f32; N] {
+        // Magic value that indicates deltas should not be applied.
+        const NO_VARIATION_DELTAS: u32 = 0xFFFFFFFF;
+        // Note: FreeType never returns an error for these lookups, so
+        // we do the same and just `unwrap_or_default` on var store
+        // errors.
+        // See <https://gitlab.freedesktop.org/freetype/freetype/-/blob/fc01e7dd/src/sfnt/ttcolr.c#L574>
+        let mut deltas = [0.0; N];
+        if self.coords.is_empty()
+            || self.var_store.is_none()
+            || var_index_base == NO_VARIATION_DELTAS
+        {
+            return deltas;
+        }
+        let var_store = self.var_store.as_ref().unwrap();
+        let compute_delta = |ty: &DeltaTy, delta_ix| {
+            match ty {
+                DeltaTy::FWord => var_store.compute_delta_f32::<FWord>(delta_ix, self.coords),
+                DeltaTy::UfWord => var_store.compute_delta_f32::<UfWord>(delta_ix, self.coords),
+                DeltaTy::Fixed => var_store.compute_delta_f32::<Fixed>(delta_ix, self.coords),
+                DeltaTy::F2Dot14 => var_store.compute_delta_f32::<F2Dot14>(delta_ix, self.coords),
+            }
+            .unwrap_or_default()
+        };
+        if let Some(index_map) = self.index_map.as_ref() {
+            for (i, (ty, delta)) in tys.iter().zip(deltas.iter_mut()).enumerate() {
+                let var_index = var_index_base + i as u32;
+                if let Ok(delta_ix) = index_map.get(var_index) {
+                    *delta = compute_delta(ty, delta_ix);
+                }
+            }
+        } else {
+            for (i, (ty, delta)) in tys.iter().zip(deltas.iter_mut()).enumerate() {
+                let var_index = var_index_base + i as u32;
+                // If we don't have a var index map, use our index as the inner
+                // component and set the outer to 0.
+                let delta_ix = DeltaSetIndex {
+                    outer: 0,
+                    inner: var_index as u16,
+                };
+                *delta = compute_delta(ty, delta_ix);
+            }
+        }
+        deltas
+    }
+}
+
+#[derive(Copy, Clone)]
+enum DeltaTy {
```

Comment:
Does `Ty` mean Type? Is that a convention we're using. Can we spell it out? `DeltaType`?

Reply:
Oops, yes. Since `type` is a reserved word in rust, this is fairly idiomatic and I follow it by habit but I’m happy to spell it out.

Reply:
Done